File: RotatedArray.java
// Time Complexity : O(log n)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

class Solution {
    public int search(int[] nums, int target) {
        int low = 0;
        int high = nums.length - 1;

        while(low <= high){
            int mid = low + (high - low) / 2;
            if (nums[mid] == target){ //If middle element is equal to target
                return mid;
            }
            if (nums[low] <= nums[mid]){ //Checks if it is left sorted
                if (nums[low] <= target && nums[mid] > target){ //Checks if target lies in the left side of the array
                    high = mid - 1; //Moves left
                }   else {
                    low = mid + 1; //Moves right
                }
            } else {                    //Checks if it is right sorted
                if (nums[mid] < target && nums[high] >= target){ //Checks if target lies in the right side of the array
                    low = mid + 1; //Moves left
                } else {
                    high = mid - 1; //Moves right
                }

            }
        }
        return -1;
    }
}

File: Rotated_array.py
#time complexity = O(logn)
# space complexity = O(1)


class Solution(object):
    def search(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: int
        """
        low,high = 0, len(nums)- 1
        
        while(low <= high):
            mid = low + (high - low) / 2    #calculate middle index
            if nums[mid] == target:  
                return mid

            if nums[low] <= nums[mid]:
                #if target in the left half
                if nums[low] <= target < nums[mid]:
                    high = mid - 1
                else:
                        low = mid +1

            else:
                #if the target in the right side
                if nums[mid] < target <= nums[high]:
                    low = mid +1
                else:
                    high = mid - 1
        #if target not found
        return -1